core/array.rbs

```diff
@@ -627,8 +628,8 @@ class Array[unchecked out Elem] < Object
   #     a #=>  ["", "b", "c!", "d!"]
   #
   # collect! is monomorphic because of RBS limitation.
-  def collect!: () { (Elem item) -> Elem } -> self
-              | () -> ::Enumerator[Elem, self]
+  def collect!: [U] () { (Elem item) -> U } -> ::Array[U]
```

Comment:
Hm, in such cases, is it really better to just return `self`?

I understand that it makes the signature describe the method behaviour better, however I think using generics here will help the signature to better describe the type behaviour. To me this is more what type signatures are for, as we can refer to the documentation or implementation if we are interested in learning the side-effects of the method.

Reply:
The following is the case to detect a type error in my mind.

```rb
a = [1,2,3]
a.map! {|x| x.to_s }
a[0] + 1
```

On the other hand, the following runs without any error == Steep detects a false error.

```rb
a = [1,2,3]
b = a.map! {|x| x.to_s }
b[0] + "string"
```

I'm not sure if we can find a solution to make the both work, and we may have to choose one. My intuition is _using `#map!` is usually done without assignment because it is in-place_.

Any thoughts from @mame regarding to TypeProf?

Reply:
Currently, TypeProf performs weak-update of the element type of a container type when its method accepts a new type at a contravariance occurrence, i.e.,

```
ary = [1, 2, 3]
ary.map!(&:to_s)
p ary #=> Array[Integer | String]
```
https://mame.github.io/typeprof-playground/#rb=ary+%3D+%5B1%2C+2%2C+3%5D%0Aary.map%21%28%26%3Ato_s%29%0Ap+ary+%23%3D%3E+Array%5BInteger+%7C+String%5D&rbs=

Changing the RBS declaration as proposed will break this behavior of TypeProf.

That being said, I have no strong opinion about the change. I'm unsure how much is used a call to `map!` that changes the element type. If there are a lot of such cases, prohibiting them could be very conservative and annoying. @hjwylde do you have any statistics?
BTW, I guess the return type of `map!` is not important. IMO it could be even `void`.

Reply:
I'm afraid I don't have any statistics. I too don't feel strongly about the return type of `map!`, so I will revert this one, and make the appropriate changes to the other points in this review - thank you.

> TypeProf performs weak-update of the element type of a container type when its method accepts a new type

This is interesting, perhaps we can adopt something similar too. Initially I had disregarded the idea as I thought it would lead to issues with `Array#include?` etc., but it seems these types of methods take an `untyped` object and it would be okay.